stake-pool/program/src/instruction.rs

```diff
@@ -0,0 +1,176 @@
+//! Instruction types
+
+#![allow(clippy::too_many_arguments)]
+
+use solana_sdk::instruction::AccountMeta;
+use solana_sdk::instruction::Instruction;
+use solana_sdk::program_error::ProgramError;
+use solana_sdk::pubkey::Pubkey;
+use std::mem::size_of;
+
+/// Fee rate as a ratio
+/// Fee is minted on deposit
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Fee {
+    /// denominator of the fee ratio
+    pub denominator: u64,
+    /// numerator of the fee ratio
+    pub numerator: u64,
+}
+
+/// Inital values for the Stake Pool
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct InitArgs {
+    /// Fee paid to the owner in pool tokens
+    pub fee: Fee,
+    /// Nonce used for the deposit program address
+    pub deposit_nonce: u8,
+    /// Nonce used for the withdraw program address
+    /// This program address is used as the stake withdraw key as well
+    pub withdraw_nonce: u8,
+}
+
+/// Instructions supported by the StakePool program.
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum StakePoolInstruction {
+    ///   Initializes a new StakePool.
+    ///
+    ///   0. `[w, s]` New StakePool to create.
+    ///   1. `[]` Owner
+    ///   2. `[]` pool token Mint. Must be non zero, owned by withdraw authority.
+    ///   3. `[]` Pool Account to deposit the generated fee for owner.
+    ///   4. `[]` Token program id
+    Initialize(InitArgs),
+
+    ///   Deposit some stake into the pool.  The output is a "pool" token representing ownership
+    ///   into the pool. Inputs are converted to the current ratio.
+    ///
+    ///   0. `[]` StakePool
+    ///   1. `[]` deposit authority
+    ///   2. `[]` withdraw  authority
+    ///   3. `[w]` Stake, deposit authority is set as the withdrawal key
+    ///   4. `[w]` Pool MINT account, authority is the owner.
+    ///   5. `[w]` Pool Account to deposit the generated tokens.
+    ///   6. `[w]` Pool Account to deposit the generated fee for owner.
+    Deposit,
+
+    ///   Withdraw the token from the pool at the current ratio.
+    ///   The amount withdrawn is the MIN(u64, stake size)
+    ///   
+    ///   0. `[]` StakePool
+    ///   1. `[]` withdraw  authority
+    ///   2. `[w]` SOURCE Pool account, amount is transferable by authority
+    ///   3. `[w]` Pool MINT account, authority is the owner
+    ///   4. `[w]` Stake SOURCE owned by the withdraw authority  
```

Comment:
Double-checking how this works -- the user needs to get all of the stake accounts controlled by the withdraw authority and choose one with enough tokens for this, correct?  That could cause some issues for people trying to withdraw a large amount, and makes me think it would be better for the StakePool to control one stake account, that it merges other stakes into.  The drawback with that approach is that then the StakePool can only delegate to one staking authority.  How do you envision this being used?

Reply:
@joncinque that's ok, they can aggregate on the UI side.